Formula/newrelic-infra-agent.rb

```diff
@@ -0,0 +1,63 @@
+class NewrelicInfraAgent < Formula
+  desc "New Relic infrastructure agent"
+  homepage "https://github.com/newrelic/infrastructure-agent"
+  url "https://github.com/newrelic/infrastructure-agent/archive/refs/tags/1.20.0.tar.gz"
+  sha256 "b9f6f744b8376b0ba92657237902d8a8e7c12cbb7f56c6f13f93b52255f31224"
+  license "Apache-2.0"
+  head "https://github.com/newrelic/infrastructure-agent.git"
+
+  depends_on "go@1.14" => :build
+
+  def plist
+    <<~EOS
+        <?xml version="1.0" encoding="UTF-8"?>
+        <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+        <plist version="1.0">
+          <dict>
+            <key>Label</key>
+            <string>com.newrelic.infrastructure.agent</string>
+            <key>Program</key>
+            <string>/usr/local/bin/newrelic-infra-service</string>
+            <key>ProgramArguments</key>
+            <array>
+              <string>/usr/local/bin/newrelic-infra-service</string>
+              <string>-config</string>
+              <string>/usr/local/etc/newrelic-infra/newrelic-infra.yml</string>
+            </array>
+            <key>RunAtLoad</key>
+            <true/>
+            <key>StandardOutPath</key>
+            <string>/usr/local/var/log/newrelic-infra/newrelic-infra.log</string>
+            <key>StandardErrorPath</key>
+            <string>/usr/local/var/log/newrelic-infra/newrelic-infra.stderr.log</string>
+            <key>ExitTimeOut</key>
+            <integer>30</integer>
+          </dict>
+        </plist>
+    EOS
+  end
+
+  def install
+    ENV["VERSION"] = version
+    ENV["GOOS"] = "darwin"
+    ENV["GOARCH"] = "amd64"
+    ENV["CGO_ENABLED"] = "1"
+    system "make", "dist-for-os"
+    on_macos do
+      bin.install "dist/darwin-newrelic-infra_darwin_amd64/newrelic-infra"
+      bin.install "dist/darwin-newrelic-infra-ctl_darwin_amd64/newrelic-infra-ctl"
+      bin.install "dist/darwin-newrelic-infra-service_darwin_amd64/newrelic-infra-service"
+    end
+
+    (etc/"newrelic-infra").mkpath
+    (var/"log/newrelic-infra").mkpath
+    (var/"db/newrelic-infra").mkpath
+
+    (var/"db/newrelic-infra").install "assets/licence/LICENSE.macos.txt"
+  end
+
+  test do
+    output = shell_output("#{bin}/newrelic-infra -version")
+    assert_match(/New\ Relic\ Infrastructure\ Agent\ version:\ 1\.20\.0/, output)
```

Comment:
We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
This app is an infra-agent of NewRelic. Usually it starts, make a connection to the NR backend and start to send host telemetry data like CPU usage, memory usage, etc. Unfortunately we  don't have this easy to test flow yet. Can we continue without self test in this case and add it later?

https://github.com/newrelic/infrastructure-agent
https://docs.newrelic.com/docs/infrastructure/infrastructure-monitoring/get-started/get-started-infrastructure-monitoring/

Reply:
> Can we continue without self test in this case and add it later?

No, the experience learns that if people don't write a good test at first they won't come back to fix it either.